File: packages/host4_flutter_aivoice/ios/AiVoice_副本/Tools/SubtitleTextAssembler.m
//
//  SubtitleTextAssembler.m
//  字幕文本拼接器实现
//

#import "SubtitleTextAssembler.h"

@interface SubtitleTextAssembler ()

// 缓存已确认的文本 [key: "userId-roundId" -> 已确认文本]
@property (nonatomic, strong) NSMutableDictionary<NSString *, NSString *> *confirmedTextCache;

// 缓存临时文本 [key: "userId-roundId" -> 临时文本]
@property (nonatomic, strong) NSMutableDictionary<NSString *, NSString *> *tempTextCache;

// 完成状态 [key: "userId-roundId" -> BOOL]
@property (nonatomic, strong) NSMutableDictionary<NSString *, NSNumber *> *completeStatusCache;

// ✅ 新增：已处理的消息集合 [key: "userId-roundId" -> Set<消息Hash>]
@property (nonatomic, strong) NSMutableDictionary<NSString *, NSMutableSet<NSString *> *> *processedMessagesCache;

@end

@implementation SubtitleTextAssembler

#pragma mark - Singleton

+ (instancetype)sharedAssembler {
    static SubtitleTextAssembler *instance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        instance = [[SubtitleTextAssembler alloc] init];
    });
    return instance;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        _confirmedTextCache = [NSMutableDictionary dictionary];
        _tempTextCache = [NSMutableDictionary dictionary];
        _completeStatusCache = [NSMutableDictionary dictionary];
        _processedMessagesCache = [NSMutableDictionary dictionary];
    }
    return self;
}

#pragma mark - Public Methods

- (NSString *)assembleText:(SubtitleMsgData *)subtitleData {
    if (!subtitleData || !subtitleData.userId) {
        return @"";
    }
    
    // 生成缓存键
    NSString *key = [self makeKey:subtitleData.userId roundId:subtitleData.roundId];
    
    // ✅ 关键修复：检查消息是否已处理
    if ([self isMessageProcessed:subtitleData forKey:key]) {
        // 已处理过，直接返回当前完整文本
        NSString *confirmedText = self.confirmedTextCache[key] ?: @"";
        NSString *tempText = self.tempTextCache[key] ?: @"";
        NSString *completeText = [confirmedText stringByAppendingString:tempText];
        
        NSLog(@"⚠️ [TextAssembler] 消息已处理，跳过: \"%@\"", subtitleData.text);
        
        return completeText;
    }
    
    // ✅ 标记消息为已处理
    [self markMessageAsProcessed:subtitleData forKey:key];
    
    // 获取已确认的文本
    NSString *confirmedText = self.confirmedTextCache[key] ?: @"";
    
    NSString *completeText = @"";
    
    if (subtitleData.definite) {
        // ===== 已确认文本 =====
        
        // 追加到已确认部分
        confirmedText = [confirmedText stringByAppendingString:subtitleData.text];
        
        // 更新缓存
        self.confirmedTextCache[key] = confirmedText;
        
        // 清空临时文本
        [self.tempTextCache removeObjectForKey:key];
        
        // 完整文本就是已确认文本
        completeText = confirmedText;
        
        // 检查是否完成
        if (subtitleData.paragraph) {
            self.completeStatusCache[key] = @(YES);
//            NSLog(@"✅ [TextAssembler] %@-%ld 完成: \"%@\"",
//                  subtitleData.userId, (long)subtitleData.roundId, completeText);
        } else {
//            NSLog(@"✅ [TextAssembler] %@-%ld 已确认: \"%@\"",
//                  subtitleData.userId, (long)subtitleData.roundId, subtitleData.text);
        }
        
    } else {
        // ===== 临时文本 =====
        
        // 保存临时文本
        self.tempTextCache[key] = subtitleData.text;
        
        // 完整文本 = 已确认 + 临时
        completeText = [confirmedText stringByAppendingString:subtitleData.text];
        
//        NSLog(@"🔄 [TextAssembler] %@-%ld 临时: \"%@\"",
//              subtitleData.userId, (long)subtitleData.roundId, subtitleData.text);
    }
    
    // 打印当前完整文本
//    NSLog(@"📄 [TextAssembler] 完整文本: \"%@\"", completeText);
    
    return completeText;
}

- (BOOL)isCompleteForUser:(NSString *)userId roundId:(NSInteger)roundId {
    NSString *key = [self makeKey:userId roundId:roundId];
    NSNumber *status = self.completeStatusCache[key];
    return status ? status.boolValue : NO;
}

- (void)clearCache {
    [self.confirmedTextCache removeAllObjects];
    [self.tempTextCache removeAllObjects];
    [self.completeStatusCache removeAllObjects];
    [self.processedMessagesCache removeAllObjects];
//    NSLog(@"🗑️ [TextAssembler] 已清空所有缓存");
}

- (void)clearCacheForUser:(NSString *)userId roundId:(NSInteger)roundId {
    NSString *key = [self makeKey:userId roundId:roundId];
    [self.confirmedTextCache removeObjectForKey:key];
    [self.tempTextCache removeObjectForKey:key];
    [self.completeStatusCache removeObjectForKey:key];
    [self.processedMessagesCache removeObjectForKey:key];
//    NSLog(@"🗑️ [TextAssembler] 已清空 %@ 的缓存", key);
}

#pragma mark - Private Methods

- (NSString *)makeKey:(NSString *)userId roundId:(NSInteger)roundId {
    return [NSString stringWithFormat:@"%@-%ld", userId, (long)roundId];
}

#pragma mark - Message Deduplication

// ✅ 新增：检查消息是否已处理
- (BOOL)isMessageProcessed:(SubtitleMsgData *)subtitleData forKey:(NSString *)key {
    // 生成消息唯一标识
    NSString *messageHash = [self generateMessageHash:subtitleData];
    
    // 获取或创建已处理消息集合
    NSMutableSet *processedSet = self.processedMessagesCache[key];
    if (!processedSet) {
        processedSet = [NSMutableSet set];
        self.processedMessagesCache[key] = processedSet;
    }
    
    // 检查是否已处理
    return [processedSet containsObject:messageHash];
}

// ✅ 新增：标记消息为已处理
- (void)markMessageAsProcessed:(SubtitleMsgData *)subtitleData forKey:(NSString *)key {
    NSString *messageHash = [self generateMessageHash:subtitleData];
    
    NSMutableSet *processedSet = self.processedMessagesCache[key];
    if (!processedSet) {
        processedSet = [NSMutableSet set];
        self.processedMessagesCache[key] = processedSet;
    }
    
    [processedSet addObject:messageHash];
}

// ✅ 新增：生成消息唯一标识
- (NSString *)generateMessageHash:(SubtitleMsgData *)subtitleData {
    // 基于内容 + 状态生成 hash
    NSString *hashString = [NSString stringWithFormat:@"%@_%d_%d",
                           subtitleData.text ?: @"",
                           subtitleData.definite,
                           subtitleData.paragraph];
    
    return @(hashString.hash).stringValue;
}

@end

File: packages/host4_flutter_aivoice/ios/Classes/Models/SubtitleTextAssembler.swift
import Foundation

/// 字幕文本拼接器 — 处理 SubtitleMsgData 并返回完整文本
class SubtitleTextAssembler {
  static let shared = SubtitleTextAssembler()

  // 缓存 key: "userId-roundId"
  private var confirmedTextCache: [String: String] = [:]
  private var tempTextCache: [String: String] = [:]
  private var completeStatusCache: [String: Bool] = [:]
  private var processedMessagesCache: [String: Set<String>] = [:]

  private init() {}

  /// 处理字幕数据，返回当前完整文本
  func assembleText(_ data: SubtitleMsgData) -> String {
    let key = "\(data.userId)-\(data.roundId)"

    // 检查是否已处理（去重）
    if isMessageProcessed(data, forKey: key) {
      let confirmed = confirmedTextCache[key] ?? ""
      let temp = tempTextCache[key] ?? ""
      return confirmed + temp
    }

    markMessageAsProcessed(data, forKey: key)

    var confirmed = confirmedTextCache[key] ?? ""

    if data.definite {
      confirmed += data.text
      confirmedTextCache[key] = confirmed
      tempTextCache.removeValue(forKey: key)

      if data.paragraph {
        completeStatusCache[key] = true
      }
      return confirmed
    } else {
      tempTextCache[key] = data.text
      return confirmed + data.text
    }
  }

  /// 检查消息是否完成
  func isCompleteFor(userId: String, roundId: Int) -> Bool {
    return completeStatusCache["\(userId)-\(roundId)"] ?? false
  }

  /// 清空所有缓存
  func clearCache() {
    confirmedTextCache.removeAll()
    tempTextCache.removeAll()
    completeStatusCache.removeAll()
    processedMessagesCache.removeAll()
  }

  /// 清空指定用户的缓存
  func clearCacheFor(userId: String, roundId: Int) {
    let key = "\(userId)-\(roundId)"
    confirmedTextCache.removeValue(forKey: key)
    tempTextCache.removeValue(forKey: key)
    completeStatusCache.removeValue(forKey: key)
    processedMessagesCache.removeValue(forKey: key)
  }

  // MARK: - Private

  private func messageHash(_ data: SubtitleMsgData) -> String {
    return "\(data.text)_\(data.definite)_\(data.paragraph)"
  }

  private func isMessageProcessed(_ data: SubtitleMsgData, forKey key: String) -> Bool {
    let hash = messageHash(data)
    return processedMessagesCache[key]?.contains(hash) ?? false
  }

  private func markMessageAsProcessed(_ data: SubtitleMsgData, forKey key: String) {
    let hash = messageHash(data)
    if processedMessagesCache[key] == nil {
      processedMessagesCache[key] = Set()
    }
    processedMessagesCache[key]?.insert(hash)
  }
}
